src/System.Diagnostics.Process/src/System/Diagnostics/Process.NonUap.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+
+namespace System.Diagnostics
+{
+    public partial class Process : IDisposable
+    {
+        public void Kill(bool entireProcessTree)
+        {
+            if (!entireProcessTree)
+            {
+                Kill();
+            }
+            else
+            {
+                EnsureState(State.Associated | State.IsLocal | State.HaveNonExitedId);
+
+                if (IsInTreeOf(GetCurrentProcess()))
+                    throw new InvalidOperationException(SR.KillEntireProcessTree_DisallowedBecauseTreeContainsCallingProcess);
+
+                IEnumerable<Exception> result = KillTree();
+
+                if (result.Any())
+                    throw new AggregateException(SR.KillEntireProcessTree_TerminationIncomplete, result);
+            }
+        }
+
+        private bool IsInTreeOf(Process processOfInterest)
+        {
+            if (SafePredicateTest(() => Equals(processOfInterest)))
+                return true;
+
+            Process[] processes = GetProcesses();
+
+            try
+            {
+                var queue = new Queue<Process>();
+                Process current = this;
+
+                do
+                {
+                    IEnumerable<Process> children = processes
```

Comment:
I might be missing something but name of this function suggests that there are 2 nodes:
- root (processOfInterest)
- potential descendant (this)
and that is suggested because the call looks like: `this.IsInTreeOf(root)` which I read as "is `this` a root or its descendant?"

The current implementation is following:
- we get a candidate which is either `this` or first element in the queue
- then a list of potential candidates is added with the children of this

```
      process of interest
          /\
         /  \
        /    \
       /\
      /  \
     /    \
          / this/current
         /
        /
       /\ next iteration (because current.IsParentOf(p) will be true)
      /  \
     /    \
```

I feel like we should be going up and next candidates should be parents (unless the function is misnamed)

am I missing anything here?

Reply:
But then on the other hand we check for `IsInTreeOf(GetCurrentProcess())` which again suggests what I wrote :-)

Reply:
> Assuming the current logic is correct, can we somehow use GetChildProcesses here?

Technically, yes. That method could be called recursively for each child process to get its children, and the on each of those children to get their children, etc. The catch is that each time it's called, it fetches a fresh copy of the currently running processes list. 

A fresh copy of that list on each call is needed in order for GetChildProcesses to satisfy the need of its current consumer. However, in this section of code, it's safe to pull that list once and then scan the cached list multiple times to build out the descendant process tree.

If we're okay with paying the (small) cost of recursively calling GetChildProcesses, I'd be glad to change to doing that.

Reply:
Could we perhaps make GetChildProcesses take a list it uses and if it is i.e. null then we fetch it?

Reply:
I won't have time to carefully look at this PR today, probably will have a bit over the weekend or Monday (ping me if I didn't)

Reply:
Reworked so that GetChildProcesses is used here.